### PR TITLE
Update TestDiscoverer to yield unique test names

### LINKED ISSUE #45: Visual Studio Text Explorer - Duplicate scenario names are not handled
When a scenario name is duplicated (either within or across feature files) the Visual Studio Test Explorer only picks up the first scenario it encounters.

Encountered in Visual Studio (2019 Community - Version 16.2.5).

The problem isn't manifested when running `dotnet test` - all scenarios with the same name run in this case.

Although scenarios with the same name may be confusing and not something we want to support, the present behaviour is also potentially confusing. This is particularly so when a scenario with a duplicated name fails. Running the tests using the VS Test Explorer appears to show that all tests are passing. Running via `dotnet test` or `dotnet vstest` (potentially on a build agent) shows the test failure.

The issue has been reproduced in this branch: https://github.com/GivePenny/GherkinSpec.SimpleExample/tree/duplicate-scenario-names

I'm happy to work on a PR to resolve this. @DaveAurionix what are your thoughts on a solution? We could fail when detecting duplicate scenario names, although the experience when this happens (for example when a feature file is not syntactically correct) isn't great. An alternative might be to append the duplicated scenario names with a counter of some sort.

### PR DESCRIPTION
**Pull request by carl-hartshorn**
Resolves #45

### CODE REVIEW

**Comment by carl-hartshorn** [comment]
I was wondering about pushing the EnsureTestCaseNameIsUnique method and the fields it relies on out to a separate class, just trying to think of a sensible name for that class.

File: GherkinSpec.TestAdapter.UnitTests/TestDiscovererShould.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        private class NullMessageLogger : IMessageLogger
+        {
+            public void SendMessage(TestMessageLevel testMessageLevel, string message)
+            {
+            }
+        }
+
+        private class TestCaseDiscoverySink : ITestCaseDiscoverySink
```
Comment (DaveAurionix) [inline GherkinSpec.TestAdapter.UnitTests/TestDiscovererShould.cs]: These are good, I can see us using them in more unit tests - maybe move them out into their own files (so one class per file)?  `NullMessageLogger` and maybe `MockTestCaseDiscoverySink`?

File: GherkinSpec.TestAdapter/TestDiscoverer.cs [hunk 2]
```diff
@@ -28,12 +32,13 @@ public class TestDiscoverer : ITestDiscoverer
...
         }
 
-        public static IEnumerable<TestCase> DiscoverTests(IEnumerable<string> sources, IMessageLogger logger)
+        public IEnumerable<TestCase> DiscoverTests(IEnumerable<string> sources, IMessageLogger logger)
         {
             foreach (var source in sources)
             {
                 foreach (var testCase in DiscoverTests(source, logger))
                 {
+                    EnsureTestCaseNameIsUnique(testCase);
```
Comment (DaveAurionix) [inline GherkinSpec.TestAdapter/TestDiscoverer.cs]: I think we may need to only ensure names are unique _within the same source_.  It adds weight to your question about whether we should move the two private fields and name-check logic into a separate class.  We could then in this method use a fresh instance for each test source?

File: GherkinSpec.TestAdapter/TestDiscoverer.cs [hunk 3]
```diff
@@ -16,6 +17,9 @@ namespace GherkinSpec.TestAdapter
     [FileExtension(".exe")]
     public class TestDiscoverer : ITestDiscoverer
     {
+        private readonly HashSet<string> discoveredFullyQualifiedTestNames = new HashSet<string>();
+        private readonly Regex uniqueTestCaseSuffixRegex = new Regex(" \\((\\d+)\\)$");
```
Comment (DaveAurionix) [inline GherkinSpec.TestAdapter/TestDiscoverer.cs]: uniqueTestCaseSuffixRegex can be static?

File: GherkinSpec.TestAdapter.UnitTests/TestDiscovererShould.cs (hunk 1 above)
Comment (carl-hartshorn) [reply GherkinSpec.TestAdapter.UnitTests/TestDiscovererShould.cs]: Thanks, extracted these to separate files.

File: GherkinSpec.TestAdapter/TestDiscoverer.cs (hunk 2 above)
Comment (carl-hartshorn) [reply GherkinSpec.TestAdapter/TestDiscoverer.cs]: Thanks, extracted this method & fields to `UniqueTestCaseNameResolver` & instantiated in the scope of each source.

File: GherkinSpec.TestAdapter/TestDiscoverer.cs (hunk 3 above)
Comment (carl-hartshorn) [reply GherkinSpec.TestAdapter/TestDiscoverer.cs]: Thanks, made this static on `UniqueTestCaseNameResolver`.

**Pull request merged**